File: src/main/java/at/jku/ssw/coco/intellij/reference/CocoCompilerReference.java
package at.jku.ssw.coco.intellij.reference;

import at.jku.ssw.coco.intellij.CocoUtil;
import at.jku.ssw.coco.intellij.psi.CocoEnd;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CocoCompilerReference extends PsiReferenceBase<CocoEnd> implements PsiReference {

    public CocoCompilerReference(@NotNull CocoEnd element, TextRange textRange) {
        super(element, textRange);
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        PsiElement ident = myElement.getIdent();
        if (ident == null) {
            return null;
        }
        return CocoUtil.findCompiler(myElement.getContainingFile(), ident.getText());
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        PsiElement resolve = resolve();
        return resolve != null ? new Object[]{resolve} : new Object[0];
    }
}

File: src/main/kotlin/me/salzinger/intellij/coco/reference/CocoCompilerReference.kt
package me.salzinger.intellij.coco.reference

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import me.salzinger.intellij.coco.findByName
import me.salzinger.intellij.coco.findCompilers
import me.salzinger.intellij.coco.psi.HasCocoCompilerReference

class CocoCompilerReference(element: HasCocoCompilerReference, textRange: TextRange) :
    AbstractRenamableReference<HasCocoCompilerReference>(element, textRange), PsiReference {

    override fun resolve(): PsiElement? {
        return findCompilers(myElement.containingFile).findByName(myElement.name)
    }

    override fun getVariants(): Array<Any> {
        return findCompilers(myElement.containingFile).toTypedArray()
    }
}
